### PR TITLE
Sort fast reindex

### PR DESCRIPTION
**Pull request by dannyroberts**
Functional change:
- Cases sort by `server_date_modified` (descending)
- Forms sort by `received_on` (descending)

Framework feature:
- an `ElasticReindexer` subclass can specify a `sort_key` function that takes in a row and returns a string; view rows will then be processed in the order specified by this sort key.
- they may also specify `sort_key_include_docs` (boolean), which if set to `True` will `include_docs` on the view for the sake of the `sort_key` function (but otherwise stripping it).
- by default (and you currently can't change the default) rows are sorted descending. (This is because our current only use case is to index most recent dates first.)

Implementation detail:
- used POSIX `sort` rather than implementing something in python because it's optimized and memory efficient for large files (i.e. I believe it does a merge sort using temp files on disk so it never has to read the entire file into memory)

### CODE REVIEW

**Comment by dannyroberts** [comment]
@esoergel @gcapalbo @czue @snopoke

File: corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py [hunk 1]
```diff
@@ -159,13 +166,35 @@ def load_from_view(self):
...
+        dump_filename = self.get_dump_filename()
+        dump_presort_filename = dump_filename + '.presort.tmp'
+        self.log('Writing dump file to disk: {}, starting at {}'.format(
+            dump_presort_filename, datetime.utcnow().isoformat()))
+        with open(dump_presort_filename, 'w') as fout:
+            for row in self.full_couch_view_iter():
+                sort_key = self.sort_key(row)
+                if self.sort_key_include_docs and not self.pillow_class.include_docs:
+                    del row['doc']
+                fout.write('{}\t{}\n'.format(sort_key, json.dumps(row)))
```
Comment (gcapalbo) [inline corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py]: if the sort key ends up with a tab in it for some reason we'd have a problem. maybe worth converting tabs to spaces in the sort_key?

Comment (gcapalbo) [inline corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py]: (and maybe convert \r and \n to spaces too)

File: corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py [hunk 2]
```diff
@@ -159,13 +166,35 @@ def load_from_view(self):
             fout.write(str(current_db_seq))
 
         # Load entire view to disk
+        if self.sort_key:
+            self._load_from_view_and_sort()
+        else:
+            self._load_from_view()
+        self.log("View and sequence written to disk: %s" % datetime.utcnow().isoformat())
+
+    def _load_from_view_and_sort(self):
```
Comment (czue) [inline corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py]: took a little while to parse through this. maybe add a doc string with something like: "writes a tab delimited file where the first key is the sort key and the second is the row (either doc or ID), then sorts by the sort key and removes it from the final output file leaving just the row" (assuming i read it right)

**Comment by czue** [comment]
looks good, thanks for the quick turnaround! :+1: on :white_check_mark:

File: corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py (hunk 1 above)
Comment (czue) [inline corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py]: @gcapalbo always with the better :eyes: for :mag: than me

**Comment by gcapalbo** [comment]
just one comment otherwise looks good from my side, also was this tested locally?

File: corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py]: Yeah, that's right. I'll leave a comment.

**Comment by dannyroberts** [comment]
@gcapalbo yup, this worked fine locally, and as far as I can tell it looked correctly sorted. I also quickly checked that it didn't obviously break fast reindexers that don't use the new feature.

File: corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py (hunk 1 above)
Comment (dannyroberts) [inline corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py]: I think I'll just assert that the key doesn't contain any `'\t\n\r'` characters—I can't imagine why they'd be there, and I'd want to be yelled at if they were.

File: corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/hqcase/management/commands/ptop_fast_reindexer.py]: Done. I may have got a little wordy, but that's the gist of it.

**Comment by esoergel** [comment]
I was just helping Will with a local elasticsearch issue and `ptop_preindex` and `run_ptop` both had cryptic failures.  `ptop_preindex` just hung forever.  We checked out a commit from a couple days ago and it worked fine.  Anyways, after this is all over we might need to revert these commits too.

**Comment by czue** [comment]
weird. i'd be interested to hear more about the ways that they failed. `preindex_everything` worked flawlessly on india, fyi.

**Comment by dannyroberts** [comment]
@esoergel that really is strange; I wouldn't expect `run_ptop` to interact with the code in this change at all. (It's for fast reindexes.)

**Comment by esoergel** [comment]
Yeah, it could've been a fluke or something up with his environment, but at least we should keep this in mind if other people report issues locally.

**Pull request merged**